feat(geometry): Adding Geometry Objects

chriswmackey opened the pull request:
You can see that this is a fairly large PR with ~5k lines of pythonic new code and docstrings.  It gives us more than half of the things originally on the checklist for this library as you can see on the [updated Readme page on my fork](https://github.com/chriswmackey/ladybug-geometry/blob/development/README.md)

If you want to get an initial sense of the capabilities, I recommend browsing the [Surface3D class](https://github.com/chriswmackey/ladybug-geometry/blob/development/ladybug_geometry/geometry3d/surface.py) since many of the other classes in the library are in service of this one and this is also the one that I have been imaging would describe honeybee surfaces, essentially allowing us to move/transform them, generate glazing based on a ratio from them, generate analysis grids from them, etc.

The main reason why this is a draft PR is that I have been doing most of the testing of the library with the help of Rhino since it is much easier to generate a lot of cases and visually evaluate them there.  So, while I am confident that 80% of the code works well from these tests, this means that I don't have any pure Python tests for CI and there are probably some bugs in the parts that have not been tested with Rhino.  So, the next step for me is writing these tests.

In the meantime, I could definitely use feedback on the code and especially the class names and organization of the library.  The organization seems very clear to me but I have a feeling this is because I have been in the zone working on it the last week.

mostaphaRoudsari commented:
Thanks @chriswmackey! This is a long one. Let's see what will be the best strategy to review it. A number of quick notes after checking a number of files.

1. Is Surface3D also a planar surface? If that is the case should we call it a `Face`?
2. I would say it's better to remove `Ladybug` from `__repr__` methods. We can just call then Surface, Mesh and so on instead of _Ladybug_ Surface, ...
3. Since the geometry instances will easily be used tens of thousands of time I think we should use `__slots__` for all of them.
4. As you are writing tests it would be great if we can also use them as sample codes to show case how this library should be used.

chriswmackey commented:
@mostaphaRoudsari ,
Thank you for the quick comments and I think we definitely want to take our time on reviewing this one.  Given our busy schedule this month, I am not planning on being ready to merge this in until May but I think that I can work on writing tests as you review the code.

To respond to your comments:
1. Are you not concerned about the confusion that this could create by calling both mesh faces and objects representing polygonal surfaces by the name `Faces`?  I guess I should have asked why you decided to call [this object within the pollination API](https://github.com/pollination/api/blob/master/spec/components/schemas/model/FaceBase.yaml) a `Face` instead of a `Surface` and what you think each of these terms means.  Actually, after reading through the [Wikipedia definiteion of surface](https://en.wikipedia.org/wiki/Surface), I think I get what you mean.  Are you thinking that surfaces can be non-planar while faces are always planar?  If so, I can get behind this name change.  I am also realizing that there is not much risk of confusion with mesh faces here since there really isn't a mesh face class in this library (I just use tuples to store the indices of each face).

2. I originally had it the way that you suggested but then I started doing tests in Rhino, where I was working with both Rhino geometries and Ladybug geometries, and I realized how easy it is to get these two confused with one another.  So I figured, if I have moments of mistaking a Rhino Point3D from a Ladybug Point3D, other people working with Ladybug API in Grasshopper are definitely going to have moments like this.  Then it crossed my mind that, if we ever have a situation where the Ladybug geometry is output from a component (perhaps for speed), it would create a lot of confusion for users to see an output just called `Mesh (vertices 4, faces 1)` and then not be able to preview the mesh in the Rhino scene or treat it like a Rhino Mesh.  So I added in the `Ladybug` in front of the `__repr__` names for all of the geometries to help me (and presumably others) more easily distinguish between them and the Rhino objects.  So my question to you would be, if not adding `Ladybug` in front of the `__repr__` name, what should we do to help users distinguish these objects from their counterparts within CAD SDKs?

3. I am already using `__slots__` on all of the small objects that get called thousands of times (points, vectors, line segments, rays) and I have already witnessed what a difference it makes in my tests.  The reason why `__slots__`  is not being used on the `Polygon` or `Surface` class is that it is set up such that most of the properties are simply `None` upon creation of the object in order to help save memory and calculation time (things like `area`, `perimeter`, `is_convex`, `triangulated_mesh` are all None to start off with).  If these properties are needed, they are calculated upon request and then cached on the object such that they don't need to be recalculated in future calls to the property.  The thing is that this memory and computation saving feature conflicts with how `__slots__` works because I end up re-defining the properties from `None` to the actual property when they are requested.  There is a way around this in that I could go back to using `__slots__` and simply not define the properties as `None` (using `hasattr` on the properties instead of asking if the property is `None`).  But this type of call would be made hundreds of times for some objects and I thought calling the `hasattr` on the properties would take longer than asking if the property is `None` (admittedly, I am not sure if this is true).  Also, using `hasattr` would prevent me from using some tricks I added in like, for mesh grids, I store the face area as a single value to save memory (since all mesh faces have the same area).

4. Will definitely do!

chriswmackey commented:
@mostaphaRoudsari ,
I am just letting you know that, after our discussions yesterday, I have addressed all of the initial comments that you brought up.

mostaphaRoudsari commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -0,0 +1,1239 @@
+        self._centroid = None
+        self._is_clockwise = None
+        self._is_convex = None
+        self._is_complex = None
```
do we have `is_complex` documented somewhere as what does it mean?

chriswmackey replied:
`_is_complex` is exposed as the `is_self_intersecting` property, meaning that the shape intersects with itself like a bow tie shape.  complex vs. simple are the mathematical terms to describe this but I exposed it as is_self_intersecting because it is a lot clearer.  I'll change the name of the hidden property to match the one that is exposed.  This will make it clearer to people editing the code.

mostaphaRoudsari commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -0,0 +1,1239 @@
+            'vertices should be a list or tuple. Got {}'.format(type(vertices))
+        assert len(vertices) >= 3, 'There must be at least 3 vertices for a Face3D.' \
+            ' Got {}'.format(len(vertices))
+        assert isinstance(vertices[0], (Point3D, Point3DImmutable)), \
```
We can probably remove the first assert and test this for all the vertices to be from type `Point3D` or `Point3DImmutable`. Then we can check the length and that will be all we need to check. Only checking the first vertices doesn't look right to me.

Also as a random comment until I get a chance to do the real review do we really need the Point3D to be `mutable`? I *feel* the default Point3D should be a mutable object and in case one needs an updated point you just create a new one.

chriswmackey replied:
This is a fair point.  I put back a check for all of the vertices and I realize that this will not be much overhead.  I will say that the main reason why I put it there is to make sure the user used Point3D objects instead of Point2D objects (which will probably be caught on the first vertex) but it is nice to rest assured knowing all vertices meet the criteria.

Now that you mention the idea of all geometry objects being immutable, I really like it.  I originally thought that there would be a use for the mutable points and vectors but, now that I see all of the higher order objects like faces and meshes have to be immutable to avoid crazy amounts of code, we might as well make all points, vectors, lines, rays, and lines immutable as well.  This will simplify the code a lot.  If you agree that this is a good idea, let me know and I will implement it.

mostaphaRoudsari replied:
I would say so. I think it will be much cleaner to do it like that. It is also one of those changes that we should do sooner than later.

chriswmackey replied:
This is done. The code's a lot cleaner for this.  Thanks for the suggestion.

mostaphaRoudsari commented on ladybug_geometry/geometry2d/mesh.py:
```diff
@@ -0,0 +1,500 @@
+        _y_dim, _num_y = Mesh2D._domain_dimensions(polygon.max.y - polygon.min.y, y_dim)
+
+        # for tolerance reasons, we shrink the x and y dimensions by a very small amount
+        # this avoids the fringe cases noted in the Polygon2d.is_point_inside description
```
Can we do this reversed and expand the polygon? In that case we will ensure the the dimensions are kept the same as input and the issue with fringe cases will also go away.

chriswmackey replied:
Let me think about it.  We might not understand the full implications of the suggestion.  For example, what point do you scale the polygon from?  Does it still avoid odd situations where you don't get the grid going all of the way to the edge when you expect it to?

chriswmackey replied:
After thinking about it, I realized that we should get the same result by scaling the polygon from the center (not the centroid).  It will take a little longer to generate this scaled polygon but we can do it that way if you want to preserve the exact grid size for the few cases it might matter.  It's still going to be subject to python floating point tolerance and the input grid size usually does not perfectly match the generated grid size unless the dimensions of the bounding rectangle are perfectly divisible by the grid size.  All of this is to say that I am not sure that it will really be that different in practice but I will implement it as you say for the few cases where it may matter.

chriswmackey replied:
Actually, it wasn't from the center that it needed to be scaled but rather it needed to be scaled from the min point and moved a very tiny distance.  Now it gives the same result with no change in grid size for tolerance.

mostaphaRoudsari commented on ladybug_geometry/geometry2d/pointvector.py:
```diff
@@ -0,0 +1,337 @@
+        d = self.magnitude
+        if d:
+            return Vector2D(self.x / d, self.y / d)
+        return self.duplicate()
```
This will only be the case if the vector is (0, 0). Right? Why do we need to treat this situation differently?

chriswmackey replied:
We want to avoid an uncaught `ZeroDivisionError`.  I guess I could just do a `try / except ZeroDivisionError` instead of an `if` statement.  Does that sounds preferable?

chriswmackey replied:
After thinking about this more, I think the try / except is better. I will implement it.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+                                 .format(len(col), len(self.faces), len(self.vertices)))
+            if isinstance(col, list):
+                col = tuple(col)
+        self._colors = col
```
When colors are None, we keep it as None. There's no good reason to double the memory taken up by the mesh for a case where we are not planning on using colors. I image that this will become particularly helpful for huge butterfly meshes. If it helps better explain this decision, I note the Rhino meshes work the same way. By default, Rhino meshes have no colors assigned to them and you have to actually generate them through an method like CreateMonotoneMesh before you can actually edit the colors.

chriswmackey replied:
When colors are None, we keep it as None. There's no good reason to double the memory taken up by the mesh for a case where we are not planning on using colors. I image that this will become particularly helpful for huge butterfly meshes. If it helps better explain this decision, I note the Rhino meshes work the same way. By default, Rhino meshes have no colors assigned to them and you have to actually generate them through an method like CreateMonotoneMesh before you can actually edit the colors.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+        return self._is_color_by_face
+
+    @is_color_by_face.setter
+    def is_color_by_face(self, by_face):
```
I have had the same thought multiple times but the main reason why I still left it exposed is that it does a little bit more than say whether the colors are by face. Even when there are no colors assigned, it actually affects the way in which the mesh is translated to Rhino (essentially noting whether the mesh is color-able by face once translated). The more that I think about it, though, the type of translation really isn't important if there are no colors involved. So I agree that we can just take the setter out.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+        _verts = tuple(pt.scale(factor, origin) for pt in self.vertices)
+        return self._mesh_scale(_verts, factor)
+
+    def scale_world_origin(self, factor):
```
I would like to keep this in here because `scale_world_origin` will be much faster when you want to scale an large model and you do not care about the origin.  This is because it completely ignores the addition and subtraction of the origin point for each vertex and essentially just multiplies the vertex xyz values by the factor.  I am thinking that it will be particularly helpful for enabling people to build Honeybee energy models in different unit systems.  Right now in legacy HB, if you build a larger multizone model in a system like feet, you can end up waiting a few minutes for it to export instead of a few seconds because of the scaling operation.  Just being able to call scale_world_origin on all of the geometry should only take a few seconds.

I guess I could do what you are suggesting by writing the `scale` method it so that there's an `if origin is None` ... only multiple the vertices by the factor.  However, this means that there's an extra `if` with each vertex, which could start to add up when you have thousands of vertices in a mesh.  IF you think the extra minute or two that it could add to scaling a large mesh doesn't matter, then I'll change it to be as you suggest.

mostaphaRoudsari replied:
> ... However, this means that there's an extra if with each vertex, which could start to add up when you have thousands of vertices in a mesh.

Not sure why you need to do it for each vertex. Can't you check it once and then have two different approaches?

if origin is None:
    # do this for all the vertices
else:
    # do something else.

This is a design decision mainly and I _think_ having one method is more elegant! Up to you.

chriswmackey replied:
I agree one method is more elegant and maybe I am making too big of a deal out of the `if` calls. If I were to do what you are proposing, we would still need to keep the `scale_world_origin` method on the Point3D object or I would have to copy the code from this method into all of the other scale methods on the other objects.  Thinking about it a bit more, maybe it's ok to have this code copied everywhere as it's only one line.  Ok. I will implement it this way.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+                        'Got {}.'.format(type(ind)))
+        self._faces = faces
+
+    def _check_face_pattern(self, pattern):
```
You are correct that it is only used for `remove_faces`.  What are you suggesting we rename the method to and what should the docstring say?

mostaphaRoudsari replied:
I added a suggestion.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+        _new_verts = []
+        _new_faces = []
+        # make a dictionary that maps old vertex indices to new ones
+        _vdict = {}
```
I have only tested it with some mesh grids of a few thousand values but those tests showed that it was much faster than any of the alternatives I could think of (definitely better than doing `index()` for every single face point).  When we get to huge butterfly meshes, we'll see if we need to look for other creative solutions.

chriswmackey commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+            if isinstance(self._face_areas, (float, int)):
+                _new_f_area = self._face_areas
+            else:
+                _f_area = []
```
Very good suggestion.  I didn't realize that I could add an `if` statement at the end of the creation of a tuple like this.  I'll implement it now and there are probably a few areas in this library where I can apply this.

chriswmackey commented on ladybug_geometry/geometry2d/_1d.py:
```diff
@@ -0,0 +1,70 @@
+
+    def ToString(self):
+        """Overwrite .NET ToString."""
+        return self.__repr__()
```
Yes, each of the objects that inherits from the object here in _1d.py has a `self.__repr__` that describes that specific object.

mostaphaRoudsari replied:
I would be conservative is such cases and add it also to baseclass so if one did subclass from this class and forgot to add `__repr__` method it doesn't create errors.

chriswmackey replied:
Fair point.  I will add a `__repr__`

chriswmackey commented on ladybug_geometry/geometry2d/_2d.py:
```diff
@@ -0,0 +1,84 @@
+    @property
+    def min(self):
+        """A Point2D for the minimum bounding rectangle vertex around this geometry."""
+        if self._min is None:
```
Yes, `self._min` gets set to `None` when you initialize the class, which happens when you initialize the Polygon2D or the Mesh2D that inherit from this class.  Because there's no `__init__` on this base class, there's no way to call any of these properties except through the classes that inherit from it.  Are you saying that we need to include an `__init__` here that we never actually use?  OR do we have to have duplicated versions of this property on each of the objects that inherit from it?

mostaphaRoudsari replied:
> Are you saying that we need to include an __init__ here that we never actually use? OR do we have to have duplicated versions of this property on each of the objects that inherit from it?

Not sure what does that mean. In most cases you call the `__init__` from the baseclass when you initiate the subclass. Why will they be duplicated? 

```python
class A(object):
    def __init__(self, min=None, max=None):
        self._min = min
        self._max = max

class B(A):
    def __init__(self, min, max, length):
        A.__init__(self, min, max)
       self._length = length
```

chriswmackey replied:
> In most cases you call the __init__ from the baseclass when you initiate the subclass.

I have not had a case where I have done this yet as I usually end up overwriting the `__init__` of the base class in the class that inherits from it since the inheriting class usually has additional properties that are not in the base class, which should usually be established in `__init__`

In any case, I can add an `__init__` to this base class in the event that someone tries to make a new class inheriting from it but doesn't overwrite `__init__`.  Because the base class here is hidden, I expect it to rarely happen but I guess it's better to be safe than sorry.

chriswmackey commented on ladybug_geometry/geometry2d/line.py:
```diff
@@ -0,0 +1,224 @@
+        """
+        return LineSegment2D(self.p.scale(factor, origin), self.v * factor)
+
+    def scale_world_origin(self, factor):
```
Yea, this is on all of the objects so I'm just going to resolve this issue and we'll discuss it in your other comment.

chriswmackey commented on ladybug_geometry/geometry2d/_2d.py:
```diff
@@ -0,0 +1,84 @@
+
+    def ToString(self):
+        """Overwrite .NET ToString."""
+        return self.__repr__()
```
If there's something to discuss here, let's just discuss it once on the other comment.  This situation happens on essentially all of the objects so I know to change it on all of them if we decide it needs to be changes.

chriswmackey commented on ladybug_geometry/geometry2d/line.py:
```diff
@@ -0,0 +1,224 @@
+        """Get Point2D values along the line that subdivide it based on input distances.
+
+        Args:
+            distances: A list of distances along the line at which to subdivide it.
```
But a single value will only subdivide it evenly if that value is perfectly divisible by the length of the line.  There are cases where we really want to only set a single number for the distance between like in this method to generate contours along a Face3D:
https://github.com/chriswmackey/ladybug-geometry/blob/development/ladybug_geometry/geometry3d/face.py#L881-L905

mostaphaRoudsari replied:
Got it! thanks.

chriswmackey commented on ladybug_geometry/geometry2d/line.py:
```diff
@@ -0,0 +1,224 @@
+            parameter: The fraction between the start and end point where the
+                desired point lies. For example, 0.5 will yield the midpoint.
+        """
+        return self.p + self.v * parameter
```
I could do that but you still get a result if the parameter is outside of 0 and 1.  It just lies on the infinite line of the LineSegment3D instead of the line segment itself.  I could imagine some situations where that result might be useful but it's not used at all by any other objects in the library.  I can add it if you think it gives too much freedom and not enough structure.

mostaphaRoudsari replied:
Should be fine as you said and it is explained in docstring.

mostaphaRoudsari commented on ladybug_geometry/_mesh.py:
```diff
@@ -0,0 +1,323 @@
+        self._faces = faces
+
+    def _check_face_pattern(self, pattern):
+        """Check any input face pattern for compatability with this mesh."""
```
```suggestion
        """Check input pattern for remove faces for compatibility with this mesh."""
```

mostaphaRoudsari commented:
PS: I suggest to squash and merge them as a single commit to end up with a cleaner git history and also make @antoinedao happy! ;)

AntoineDao commented:
Realistically what would make me happy is smaller PRs merged more often rather than one spanning from end of March... But of well, you can't have it all I guess 😛

AntoineDao commented:
With regards to commit messages it's mostly for the sake of:
1. Commit history review
2. Semantic release version upgrades
3. Changelog of semantic releases being meaningful

chriswmackey commented:
Definitely going to Squash and merge this one.  The only reason why I didn't do my semantic duty is because this is the first PR that is actually establishing the repo (all future PRs will be much more diligent).  We definitely got carried away leaving this PR for 2 months and this should also not happen in the future.
